### PR TITLE
add custom formatter

### PR DESCRIPTION
**Pull request by avadan1**
[VSRE-790]

This PR adds a custom log formatter that supports adding additional tags to all logs produced by an application. See [RFC](https://vendasta.jira.com/wiki/spaces/RD/pages/2305917321/Log+correlation+with+OpenTelemetry+in+the+Yesware+Stack) for context.

@vendasta/sre 
@vendasta/meerkats 
@rafalyesware 
@faustus7 
@ancharon 

[VSRE-790]: https://vendasta.jira.com/browse/VSRE-790?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by avadan1 — COMMENTED** [review]
Correct me if I'm wrong, but it looks like all changes to YW libraries require new github releases after they're merged

File: lib/yeti_logger/custom_formatter.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  def add_tags(tags)
+    @tags.merge!(tags)
+  end
+
+  # @param severity [String]
+  # @param time [Time] unused
+  # @param progname [String] unused
+  # @param msg [String]
+  def call(severity, time, progname, msg)
+    timestamp = "#{Time.now.iso8601(3)}"
```
Comment (mvegac) [inline lib/yeti_logger/custom_formatter.rb]: Should we use UTC? 
```suggestion
    timestamp = Time.now.utc.iso8601(3)
```

File: lib/yeti_logger/custom_formatter.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    pid = Process.pid
+    msg = msg.inspect unless msg.is_a?(String)
+    msg = "#{msg}\n" unless msg[-1] == ?\n
```
Comment (mvegac) [inline lib/yeti_logger/custom_formatter.rb:24]:
```suggestion
    msg = "#{msg.inspect}\n" unless msg.is_a?(String) && msg.end_with?("\n")
```

File: lib/yeti_logger/custom_formatter.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    log_str = "#{timestamp} pid=#{pid} "
+    @tags.to_h.each do |k, v|
+      tag_val = v.call
+      if !tag_val.blank?
+        log_str += "#{k}=#{tag_val} "
+      end
+    end
```
Comment (mvegac) [inline lib/yeti_logger/custom_formatter.rb]: Avoiding checking blank will make this work in non-rails environments
```suggestion
    # Adding tags to the log string
    tag_str = @tags.map { |k, v| 
      value = v.call
      "#{k}=#{value}" unless value.to_s.empty?
    }.compact.join(' ')
```

File: lib/yeti_logger/custom_formatter.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    log_str + "[#{severity}] - #{msg}"
```
Comment (mvegac) [inline lib/yeti_logger/custom_formatter.rb]:
```suggestion
    "#{log_str}#{tag_str.empty? ? '' : ' ' + tag_str} [#{severity}] - #{msg}"
```

**Comment by mvegac** [comment]
I would also suggest to add some comments to what the logger is doing

File: lib/yeti_logger/custom_formatter.rb (hunk 1 above)
Comment (avadan1) [reply lib/yeti_logger/custom_formatter.rb]: Great catch. UTC would align with the format currently used by application logs

Screenshot 2024-03-01 at 4 09 17 PM

File: lib/yeti_logger/custom_formatter.rb (hunk 2 above)
Comment (avadan1) [reply lib/yeti_logger/custom_formatter.rb:24]: Turns out your suggestion is not equivalent with the current code.

Below are logs produced by the current code. The contents of `msg` are to the right of the dash (`-`) character.
Screenshot 2024-03-01 at 4 17 42 PM

And below are logs produced by your suggestion.
Screenshot 2024-03-01 at 4 18 46 PM

Note the additional quotation marks around the body of the log using your suggested approach.

**Review by mvegac — APPROVED** [review]
Thanks for adding the extra comments and adding some of the tweaks!

**Comment by avadan1** [comment]
> Thanks for adding the extra comments and adding some of the tweaks!

Thank you for the input 🚀

**Pull request merged**